### PR TITLE
py/modmath: Add math.factorial

### PR DESCRIPTION
**Pull request by swenson**
This provides a basic, mildly optimized factorial function.

There are some more optimizations we could do, but they tend to take more code, and more storage space. This seems like a sensible compromise.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for the contribution.

In embedded systems it's a good idea to try and avoid recursion whenever possible.  Did you benchmark this approach against the naive multiplication loop?  And how about looking at the "squared difference" algorithm which has half as many multiplications and is purely iterative; see eg http://www.luschny.de/math/factorial/index.html

**Comment by swenson** [comment]
I have benchmarked it in the past for other implementations, and the recursive "pyramid" approach tends to be significantly faster than the naive iterative loop for factorials where the product spans more than a few machine words. It comes down to the pyramid approach being able to take advantage of the Karatsuba implementation for the largest multiplies, whereas the iterative approach does O(N) multiplications that are O(N) in size.

(I also wrote [Guava's BigIntegerMath.factorial](https://github.com/google/guava/blob/master/guava/src/com/google/common/math/BigIntegerMath.java#L323-L409), which uses a similar approach, though I added a few optimizations there.)

Funnily enough, the "squared difference" approach is also recursive, it just hides the recursion in the call to `Xint.product`, whereas we do a simpler recursive way here.

It is possible to make the product iterative instead of recursive, but I don't think the trade-off is worth it. Notably, if the stack gets deep enough for it to matter, the size of the numbers being multiplied will far outweigh the height of the call stack.

There are definitely a lot of games we can play with trying to tune speed vs. memory vs. code simplicity. I thought it best to get a relatively simple implementation that would perform well for small-to-moderate-sized factorials, and not have embarrassing performance. :)

If you want to go deep down the factorial rabbit hole though, I'm happy to go along.

**Comment by swenson** [comment]
Ah, you're correct. I was looking at the squared difference prod method. The normal squared difference method is iterative but won't take advantage of Karatsuba for larger sizes.I can whip up some benchmarks for sure. I have a MicroPython board laying around that I should be able to use as well as my desktop.This will probably take a few days.On Sep 10, 2018 20:25, Damien George <notifications@github.com> wrote:
Funnily enough, the "squared difference" approach is also recursive, it just hides the recursion in the call to Xint.product, whereas we do a simpler recursive way here.

I was looking at the "squared difference" algo, not "squared difference product", and the former does seem to be purely iterative and not use Xint.product... or maybe I miss something?

There are definitely a lot of games we can play with trying to tune speed vs. memory vs. code simplicity. I thought it best to get a relatively simple implementation that would perform well for small-to-moderate-sized factorials, and not have embarrassing performance. :)

Yes I think this is a good approach.  But if the code is going to be optimised, even a bit, then it'd be worth doing actual benchmarks to see what is a good compromise.  (Eg the underlying bignum library we use might have different properties than standard ones so might lead to unexpected performance results.)
@swenson if you have time it would be nice to see implementations and benchmark results for 1) naive multiplication loop; 2) the approach here; 3) non-recursive squared difference.  If you want to provide the implementations I can do some benchmarking on MCU hardware, which gives very stable performance results (unlike running on a desktop).

**Comment by swenson** [comment]
Okay. I added all three versions into the code base, and uploaded the firmware to my v1.1 PyBoard.

I ran the following benchmark code on it:

```python
import math
import time


ITERS = 10
NUM = 1000

def run(f):
    t = time.ticks_us()
    f(ITERS)
    t = time.ticks_us() - t
    print("%.6f" % (t / 1e6 / float(ITERS)))

def factorial(n):
    for _ in range(n):
        math.factorial(NUM)

def factorial_recursive(n):
    for _ in range(n):
        math.factorial_recursive(NUM)

def factorial_sqdiff(n):
    for _ in range(n):
        math.factorial_sqdiff(NUM)

for NUM in [10, 14, 15, 16, 17]:
    print('factorial', NUM)
    run(factorial)
    print('factorial_recursive', NUM)
    run(factorial_recursive)
    print('factorial_sqdiff', NUM)
    run(factorial_sqdiff)
    print('')
```

The output was:

```
factorial 10
0.000038
factorial_recursive 10
0.000044
factorial_sqdiff 10
0.000036

factorial 15
0.000067
factorial_recursive 15
0.000054
factorial_sqdiff 15
0.000062

factorial 100
0.001021
factorial_recursive 100
0.000437
factorial_sqdiff 100
0.000541

factorial 1000
0.103114
factorial_recursive 1000
0.026344
factorial_sqdiff 1000
0.069515

factorial 5000
2.256677
factorial_recursive 5000
0.823874
factorial_sqdiff 5000
1.834232
```

tl;dr:

* Naive factorial is never fastest (even when checking 1!, 2!, etc.)
* Squared difference factorial is fastest up to 14!
* The recursive method is faster for >= 15.

**Comment by swenson** [comment]
(I'll delete the sqdiff and naive methods and rebase it all up if you approve @dpgeorge.)

**Comment by dpgeorge** [comment]
@swenson thanks for the detailed benchmarking.  You were right that the recursive method is faster.  But as I mentioned, recursion is not good in an embedded environment.

> Notably, if the stack gets deep enough for it to matter, the size of the numbers being multiplied will far outweigh the height of the call stack.

It looks like stack usage will go like `log2(log2(input_n))`, which is log2 of the bit width of `mp_int_t`, so 5 or 6 (for 32 or 64 bit words).  That might be acceptable, depending on how much stack a given system has.

It would also be good to compare the code size usage difference between recursive and sqdiff.

**Comment by swenson** [comment]
Here's the `stm32` assembly code for the squared difference method. It looks to be 70 bytes long.

```assembly

00000000 <mp_math_factorial_sqdiff>:
   0:   b570            push    {r4, r5, r6, lr}
   2:   f7ff fffe       bl      0 <mp_obj_get_int>
   6:   1e04            subs    r4, r0, #0
   8:   da01            bge.n   e <mp_math_factorial_sqdiff+0xe>
   a:   f7ff fffe       bl      0 <mp_math_factorial_sqdiff>
   e:   2c01            cmp     r4, #1
  10:   dd17            ble.n   42 <mp_math_factorial_sqdiff+0x42>
  12:   1065            asrs    r5, r4, #1
  14:   436d            muls    r5, r5
  16:   0069            lsls    r1, r5, #1
  18:   07e3            lsls    r3, r4, #31
  1a:   bf48            it      mi
  1c:   4361            mulmi   r1, r4
  1e:   0049            lsls    r1, r1, #1
  20:   f041 0001       orr.w   r0, r1, #1
  24:   2601            movs    r6, #1
  26:   3c02            subs    r4, #2
  28:   42b4            cmp     r4, r6
  2a:   dc00            bgt.n   2e <mp_math_factorial_sqdiff+0x2e>
  2c:   bd70            pop     {r4, r5, r6, pc}
  2e:   1bad            subs    r5, r5, r6
  30:   006a            lsls    r2, r5, #1
  32:   4601            mov     r1, r0
  34:   f042 0201       orr.w   r2, r2, #1
  38:   201c            movs    r0, #28
  3a:   f7ff fffe       bl      0 <mp_binary_op>
  3e:   3602            adds    r6, #2
  40:   e7f2            b.n     28 <mp_math_factorial_sqdiff+0x28>
  42:   2003            movs    r0, #3
  44:   e7f2            b.n     2c <mp_math_factorial_sqdiff+0x2c>
```

And for the recursive version is 134 bytes long.

```assembly
00000000 <mp_math_factorial_recursive>:
   0:   b508            push    {r3, lr}
   2:   f7ff fffe       bl      0 <mp_obj_get_int>
   6:   1e01            subs    r1, r0, #0
   8:   da01            bge.n   e <mp_math_factorial_recursive+0xe>
   a:   f7ff fffe       bl      0 <mp_math_factorial_recursive>
   e:   d004            beq.n   1a <mp_math_factorial_recursive+0x1a>
  10:   2001            movs    r0, #1
  12:   e8bd 4008       ldmia.w sp!, {r3, lr}
  16:   f7ff bffe       b.w     0 <mp_math_factorial_recursive>
  1a:   2003            movs    r0, #3
  1c:   bd08            pop     {r3, pc}

00000000 <mp_math_factorial_inner>:
   0:   4288            cmp     r0, r1
   2:   b570            push    {r4, r5, r6, lr}
   4:   460d            mov     r5, r1
   6:   d103            bne.n   10 <mp_math_factorial_inner+0x10>
   8:   e8bd 4070       ldmia.w sp!, {r4, r5, r6, lr}
   c:   f7ff bffe       b.w     0 <mp_obj_new_int>
  10:   1a0c            subs    r4, r1, r0
  12:   2c01            cmp     r4, #1
  14:   d10a            bne.n   2c <mp_math_factorial_inner+0x2c>
  16:   004a            lsls    r2, r1, #1
  18:   0041            lsls    r1, r0, #1
  1a:   f042 0201       orr.w   r2, r2, #1
  1e:   f041 0101       orr.w   r1, r1, #1
  22:   201c            movs    r0, #28
  24:   e8bd 4070       ldmia.w sp!, {r4, r5, r6, lr}
  28:   f7ff bffe       b.w     0 <mp_binary_op>
  2c:   2c02            cmp     r4, #2
  2e:   d10e            bne.n   4e <mp_math_factorial_inner+0x4e>
  30:   1c42            adds    r2, r0, #1
  32:   0041            lsls    r1, r0, #1
  34:   0052            lsls    r2, r2, #1
  36:   f042 0201       orr.w   r2, r2, #1
  3a:   f041 0101       orr.w   r1, r1, #1
  3e:   201c            movs    r0, #28
  40:   f7ff fffe       bl      0 <mp_binary_op>
  44:   006a            lsls    r2, r5, #1
  46:   f042 0201       orr.w   r2, r2, #1
  4a:   4601            mov     r1, r0
  4c:   e7e9            b.n     22 <mp_math_factorial_inner+0x22>
  4e:   eb00 0454       add.w   r4, r0, r4, lsr #1
  52:   4621            mov     r1, r4
  54:   f7ff ffd4       bl      0 <mp_math_factorial_inner>
  58:   4629            mov     r1, r5
  5a:   4606            mov     r6, r0
  5c:   1c60            adds    r0, r4, #1
  5e:   f7ff ffcf       bl      0 <mp_math_factorial_inner>
  62:   4631            mov     r1, r6
  64:   4602            mov     r2, r0
  66:   e7dc            b.n     22 <mp_math_factorial_inner+0x22>
```

**Comment by dpgeorge** [comment]
@swenson thanks again for the details.  The recursive version uses 16 bytes of stack for each call, so maybe 100 bytes total for the worst case input value.  This is not too bad.  The sqdiff version is quite a bit shorter in code size though.

So it's a trade off between speed, code size and stack usage.  I actually think it would be good to have it configurable: the sqdiff for cases where ROM is at a premium, or where recursion is not allowed.  And the recursive implementation for cases where speed is important.

**Comment by swenson** [comment]
Makes a lot of sense that it should be configurable. I see that `-Os` is the default optimization level, so `sqdiff` makes sense. Is there a particular `#ifdef` we should put the recursive version behind? I didn't see anything obvious in the source (and there seems to be no way to check the optimization level using `#idfed` sadly).

**Comment by pfalcon** [comment]
@swenson : You should put the entire behind MICROPY_PY_MATH_FACTORIAL, default off. I don't see it fitting under existing MICROPY_PY_MATH_SPECIAL_FUNCTIONS, giving that Unix port enables MICROPY_PY_MATH_SPECIAL_FUNCTIONS, while it can keep going along without math.factorial(), given how trivial it's to implement it in Python.

You should also put the test into a separate file, to not fail the entire test module when factorial is not available (and then this test should skip the test if the function not enabled). Finally, please consider runtime of calculating all factorials from 1 to 100 on slow hardware, like 8MHz CPU. If it takes more than 0.1s in this case, please reduce the upper bound, to not affect runtime of the entire testsuite.

**Comment by swenson** [comment]
Updated. Let me know if you want me to squash the commits.

File: ports/stm32/mpconfigport.h [hunk 1]
```diff
@@ -52,6 +52,7 @@
 #define MICROPY_OPT_COMPUTED_GOTO   (1)
 #define MICROPY_OPT_CACHE_MAP_LOOKUP_IN_BYTECODE (0)
 #define MICROPY_OPT_MPZ_BITWISE     (1)
+#define MICROPY_OPT_FAST_FACTORIAL  (1)
```
Comment (dpgeorge) [inline ports/stm32/mpconfigport.h]: It would be clearer to call this `MICROPY_OPT_MATH_FACTORIAL`, since the word "OPT" already means that enabling this will optimise the feature.

File: tests/float/math_fun_intbig.py [hunk 2]
```diff
@@ -9,3 +9,11 @@
 for fun in (math.ceil, math.floor, math.trunc):
     for x in (-1e25, 1e25):
         print('%.3g' % fun(x))
+
+if 'factorial' in dir(math):
```
Comment (dpgeorge) [inline tests/float/math_fun_intbig.py]: This will need to go in a separate file called `math_factorial_intbig.py`, with a feature test something like this:
```python
try:
    import math
    math.factorial
except (ImportError, NameError):
    print('SKIP')
    SystemExit
```
Then Travis CI should pass.

**Review by dpgeorge — COMMENTED** [review]
Thanks @swenson for updating.  There are some minor comments to address.

**Comment by swenson** [comment]
Updated. Looks like the coverage is still unhappy (I'm guessing because we have two options, and are only testing one), and we are increasing the size of the `stm32` port if we include factorial by default.

**Comment by dpgeorge** [comment]
> Looks like the coverage is still unhappy

It's because there is no test for math.factorial and this function is enabled in the coverage build.  Did you forget to add a file?

Minimal build also increases in size by 4 bytes, but that seems spurious.

@swenson I can take it from here if you like, unless you want to add a test file.

File: py/modmath.c [hunk 3]
```diff
@@ -232,6 +232,66 @@ STATIC mp_obj_t mp_math_degrees(mp_obj_t x_obj) {
...
+        mp_uint_t middle = start + ((end - start) >> 1);
+        mp_obj_t left = mp_math_factorial_inner(start, middle);
+        mp_obj_t right = mp_math_factorial_inner(middle + 1, end);
+        return mp_binary_op(MP_BINARY_OP_MULTIPLY, left, right);
+    }
+}
+STATIC mp_obj_t mp_math_factorial(mp_obj_t x_obj) {
+    mp_int_t max = mp_obj_get_int(x_obj);
+    if (max < 0) {
+        mp_raise_msg(&mp_type_ValueError, "factorial() not defined for negative values");
```
Comment (dpgeorge) [inline py/modmath.c]: One more thing: error messages don't need to match CPython, and this one is a bit long, something like "negative factorial" would be enough (similar to "divide by zero").

**Comment by dpgeorge** [comment]
Squashed and merged in 8c656754aa2892cbd36968bfaab1ff7033edeb0f with some minor edits to comments and whitespace.

Thanks @swenson for the contribution!
